core/src/main/scala/kafka/zk/KafkaZkClient.scala

```diff
@@ -86,14 +89,78 @@ class KafkaZkClient private (zooKeeperClient: ZooKeeperClient, isSecure: Boolean
   }
 
   /**
-   * Registers a given broker in zookeeper as the controller.
+   * Registers a given broker in zookeeper as the controller and increments controller epoch.
+   * @return the (updated controller epoch, epoch zkVersion) tuple
    * @param controllerId the id of the broker that is to be registered as the controller.
-   * @param timestamp the timestamp of the controller election.
-   * @throws KeeperException if an error is returned by ZooKeeper.
+   * @throws ControllerMovedException if fail to create /controller or fail to increment controller epoch.
    */
-  def registerController(controllerId: Int, timestamp: Long): Unit = {
-    val path = ControllerZNode.path
-    checkedEphemeralCreate(path, ControllerZNode.encode(controllerId, timestamp))
+  def registerControllerAndIncrementControllerEpoch(controllerId: Int): (Int, Int) = {
+    val timestamp = time.milliseconds()
+
+    // Create /controller_epoch if not exists
+    maybeCreateControllerEpochZNode()
+
+    // Read /controller_epoch to get the current controller epoch and zkVersion
+    val (curEpoch, curEpochStat) = getControllerEpoch.get
+
+    // Create /controller and update /controller_epoch atomically
+    val newControllerEpoch = curEpoch + 1
+    val expectedControllerEpochZkVersion = curEpochStat.getVersion
+
+    def controllerNodeExistsHandler(): (Int, Int) = {
+      val getDataRequest = GetDataRequest(ControllerZNode.path)
+      val getDataResponse = retryRequestUntilConnected(getDataRequest)
+      getDataResponse.resultCode match {
+        case Code.OK if getDataResponse.stat.getEphemeralOwner != zooKeeperClient.sessionId =>
+          error(s"Error while creating ephemeral at ${ControllerZNode.path}, node already exists and owner " +
+            s"'${getDataResponse.stat.getEphemeralOwner}' does not match current session '${zooKeeperClient.sessionId}'")
+          throw new ControllerMovedException("Controller moved to another broker. Aborting controller startup procedure")
+        case code@Code.OK =>
+          getControllerEpoch match {
+            case Some((epoch, stat)) if epoch == newControllerEpoch =>
```

Comment:
It seems that we can enter this state if broker executes `registerControllerAndIncrementControllerEpoch()` and finds that the controller znode has already been created by itself. Is this state possible?

Previously if broker tries to create controller znode and node already exists, the broker will simply read the controller id from the controller znode and move on. This patches added quite a few new logic in `controllerNodeExistsHandler()`, e.g. uses zk session id to detect whether the controller znode is created by this broker, handles the scenario that the controller znode is created by this broker. So the new code is more complicated than the previous version. Can you explain a bit why we need these new logic?

Reply:
In short, the purpose of `controllerNodeExistsHandler` is to mimic `checkedEphemeralCreate `, which previously is used to create `/controller` ephemeral node. In `CheckedEphemeral`, we need to double check the owner of the node if we saw `Code.NODEEXISTS`.

I think the purpose of the check and the additional logic is to handle transient network connection loss while creating the ephemeral.  Let's say our client sent a `create` request to zookeeper to create ephemeral znode and zookeeper receives this request and successfully creates the znode but fail to send back the response to our client because of transient network issue. In `retryUntilConnected`, our client tries to resend the request and gets the `Code.NODEEXISTS`. In this case, our client actually successfully creates and owns the znode.

Reply:
The creation of the ephemeral znode `/controller` is probably a bit different from the creation of other ephemeral znode. The broker which creates the ephemeral znode `/controller` is explicitly specified in the znode data. Thus the old approach, which reads the broker id from the controller znode after seeing `Code.NODEEXISTS`, seems OK. And that old approach seems to handle the network connection loss scenario described here. I am wondering if we can use the old approach since its logic looks simpler. What do you think?

Reply:
BTW, for the case `getDataResponse.stat.getEphemeralOwner != zooKeeperClient.sessionId`, we know this case may happen and the code will automatically recover from this. In this case it is probably better to log at warning level instead of error level. Here is a good explanation for how to choose log level. https://stackoverflow.com/questions/2031163/when-to-use-the-different-log-levels

Reply:
I am a little bit confused about what do you mean by the old approach. Are you referring to `checkedEphemeralCreate`? The logic in `controllerNodeExistsHandler ` is essentially the same as `checkedEphmeralCreate` when the node already exists, except that `controllerNodeExistsHandler` will read `/controller_epoch` to get back the epoch zkVersion when the owner of `/controller` is the current broker.

Reply:
My bad. I missed the fact that the controller znode was created using `KafkaZkClient.checkedEphemeralCreate()` which has the logic similar to what you are doing here.

It seems that the most important logic in the `KafkaZkClient.checkedEphemeralCreate` is to translate `Code.NODEEXISTS` to `Code.OK` for the znode creation operation if `getDataResponse.stat.getEphemeralOwner == zooKeeperClient.sessionId`. This logic was added in https://github.com/apache/kafka/pull/3765 by Onur. My understanding is that, in case of connection issue between broker and zookeeper, it is possible for controller znode to be successfully created and yet the return code is `Code.NODEEXISTS`. `KafkaZkClient.checkedEphemeralCreate` will handle this scenario properly. It will be good for @onurkaraman to clarify whether this understanding is correct so that we can decide whether we should keep this logic.

Here is another question. With the current patch, if the controller znode creation has failed due to znode exists exception and then broker find that `getDataResponse.stat.getEphemeralOwner == zooKeeperClient.sessionId`, it seems `registerControllerAndIncrementControllerEpoch()` can return `(newControllerEpoch, stat.getVersion)` if `if epoch == newControllerEpoch`. But is controller epoch incremented in this case? If not, then it seems something is wrong?

Reply:
I re-think about your previous suggestion for checking the payload of `/controller` only and I think it will work. I will check with Onur offline to understand more about  `checkedEphemeralCreate` and confirm.

In terms of your second concern, if we already see `getDataResponse.stat.getEphemeralOwner == zooKeeperClient.sessionId`, that means `/controller` has been created successfully. Since the only code path to create `/controller` is within a zookeeper transaction along with the `/controller_epoch` update, we can infer that the controller epoch must get incremented in this case.

Reply:
Discussed with Onur offline, the purpose of `getAfterNodeExists` in `CheckedEphemeral` is indeed used to handle the case when zk connection loss happens. After digging around both zookeeper and kafka codes, we think it is safe to remove the extra complexity for `controllerNodeExistsHandler` in this PR when we make `/controller` creation and `/controller_epoch` update atomic.

So the logic will be:
1). Try to create `/controller_epoch` if not exists
2). Read `/controller_epoch` from zk
3). Atomically create `/controller` and update `/controller_epoch`
4). If 3) throws NodeExistsException, read `/controller` and if controller id in zk equals the current broker id and if controller epoch in zk equals the expected epoch, successfully finish controller election; Otherwise, throw ControllerMovedException.